src/particles/elements/SoftQuad.H

```diff
@@ -0,0 +1,479 @@
+            SoftQuadrupole_device_copyable(other.m_gscale, other.m_mapsteps)
+        {
+#if !AMREX_DEVICE_COMPILE
+            // copy the data container if we copy the host element
+            m_cos_coef = other.m_cos_coef;
+            m_sin_coef = other.m_sin_coef;
+            amrex::Gpu::synchronize();
+#endif
+            m_ncoef = m_cos_coef.size();
+            m_cos_data = m_cos_coef.data();
+            m_sin_data = m_sin_coef.data();
+        }
+        SoftQuadrupole& operator=(SoftQuadrupole const& other)
+        {
+            if (this == &other)
+                return *this;
+
+            Thick::operator=(other);
+            SoftQuadrupole_device_copyable::operator=(other);
+#if !AMREX_DEVICE_COMPILE
+            // copy the data container if we copy the host element
+            m_cos_coef = other.m_cos_coef;
+            m_sin_coef = other.m_sin_coef;
+            amrex::Gpu::synchronize();
+#endif
+            m_ncoef = m_cos_coef.size();
+            m_cos_data = m_cos_coef.data();
+            m_sin_data = m_sin_coef.data();
+            return *this;
+        }
+        SoftQuadrupole(SoftQuadrupole && other) = default;
+        SoftQuadrupole& operator=(SoftQuadrupole && other) = default;
+
+        AMREX_GPU_HOST
+        ~SoftQuadrupole() = default;
+
+        /** Push all particles */
+        using BeamOptic::operator();
+
+        /** This is a soft-edge quadrupole functor, so that a variable of this type can be used
+         *  like a soft-edge quadrupole function.
+         *
+         * @param p Particle AoS data for positions and cpu/id
+         * @param px particle momentum in x
+         * @param py particle momentum in y
+         * @param pt particle momentum in t
+         * @param refpart reference particle
+         */
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+        void operator() (
+            PType& AMREX_RESTRICT p,
+            amrex::ParticleReal & AMREX_RESTRICT px,
+            amrex::ParticleReal & AMREX_RESTRICT py,
+            amrex::ParticleReal & AMREX_RESTRICT pt,
+            [[maybe_unused]] RefPart const & refpart
+        ) const
+        {
+            using namespace amrex::literals; // for _rt and _prt
+
+            // access AoS data such as positions and cpu/id
+            amrex::ParticleReal const x = p.pos(0);
+            amrex::ParticleReal const y = p.pos(1);
+            amrex::ParticleReal const t = p.pos(2);
+
+            // initialize output values of momenta
+            amrex::ParticleReal pxout = px;
+            amrex::ParticleReal pyout = py;
+            amrex::ParticleReal ptout = pt;
+
+            // get the linear map
+            amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const R = refpart.map;
+
+            // symplectic linear map for a quadrupole is computed using the
+            // Hamiltonian formalism as described in:
+            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf.
```

Comment:
Sometimes safer for links that people might copy:
```suggestion
            // https://uspas.fnal.gov/materials/09UNM/ComputationalMethods.pdf .
```